### DISCUSSION TITLE
How can I wait for a resource to be available

### DISCUSSION THREAD

**Discussion by badcodecat** [bevyengine/bevy#9181] [Q&A] [answered]
### My situation

I have a system that will create a resource, and a system that will use that resource to create some entities at the start of the level. Here's an example of what I have so far:
A resource:
```
#[derive(Resource)]
pub struct EnemyColour(pub Color);
```
A system to create that resource:
`fn create_resource()`

A system that will use that resource to spawn some entities at the start of the level:
`fn spawn_entities(enemy_colour: Res<EnemyColour>, ...)`

Have an ApplicationState:
```
#[derive(States, ...)]
pub enum ApplicationState
{
	#[default]
	MainMenu,
	Game,
}
```

Add systems to the app:
```
app
	.add_systems
	(
		OnEnter(ApplicationState::Game),
		create_resource
	);
```
And (separate file)
```
app
	.add_systems
	(
		OnEnter(ApplicationState::Game),
		spawn_entities
	);
```

### What I've tried

**Use `.run_if(resource_added::<EnemyColour>())` or `.run_if(resource_exists::<EnemyColour>())` on `create_entities`**

When I tried this the `create_entities` system didn't run.

**Use SystemSet**

Create a LoadingSet:
```
#[derive(SystemSet, ...)]
pub enum LoadingSet
{
	#[default]
	Loading,
	Loaded,
}
```

Configure sets:
```
app
	.configure_sets(OnEnter(ApplicationState::Game), (LoadingSet::Loading, LoadingSet::Loaded).chain())
```

Add system to the set:
```
app
	.add_systems
	(
		OnEnter(ApplicationState::Game),
		create_resource.in_set(LoadingSet::Loading)
	);
```
```
app
	.add_systems
	(
		OnEnter(ApplicationState::Game),
		spawn_entities.in_set(LoadingSet::Loaded)
	);
```
Same result, the `create_entities` system didn't run.

**Make the LoadingSet a `States`**

Convert to states:
```
#[derive(States, ...)]
pub enum LoadingState
{
	#[default]
	Loading,
	Loaded,
}
```
Try to use both states (compile error):
```
app
	.add_systems
	(
		(OnEnter(ApplicationState::Game, OnEnter(LoadingState::Loaded))),
		spawn_entities
	);
```

I got a compilation error trying to do this. I'm not sure if this method is impossible or just needs correct syntax.

Any help is appreciated, thank you for reading.

**Comment by Pieresqi** [comment]
Hi, IMO the simplest thing would be to derive Default (if your res init is simple) or FromWorld (if your res needs another res or needs to set up something else - another res, entity, etc).

And then you would use it `app.init_resource::<EnemyColour>()`

**Comment by NiklasEi** [comment, marked as answer]
`OnEnter(State)` is a schedule that only gets executed once when you change states. That means you cannot wait by letting frames run until the resource is there.
Doing anything via commands means that the action is deferred until the system `apply_deferred` is called the next time. Bevy calls this system for you in some places which explains why waiting a frame works for systems that run every frame.

Your options:
* Add the resource to the world directly in an exclusive system
* Order your systems and manually add `apply_deferred` between them
* Add the resource to the app in the very beginning via `init_resource` or `insert_resource`

**Reply by badcodecat** [reply]
`apply_deferred` worked perfectly, thanks!
